### DISCUSSION TITLE
How to load a config with FrameworkConfig $framework from a bundle

### DISCUSSION THREAD

**Discussion by jfsenechal** [symfony/symfony#42450] [Q&A] [answered]
Hello

How can I load from my bundle extension a config file who use Frameworkconfig 

```php
return static function (FrameworkConfig $framework) {
    $messenger = $framework->messenger();
    $messenger
        ->transport('async', [
            'dsn' => '%env(MESSENGER_TRANSPORT_DSN)%',
            'options' => [
                'auto_setup' => true,
                'use_notify' => true,//PostgreSQL’s
                'check_delayed_interval' => 60000,
            ],
            'retry_strategy' => [
                'max_retries' => 3,
                'multiplier' => 2,
            ],
        ]);
    $messenger->routing('Symfony\Component\Mailer\Messenger\SendEmailMessage')->senders(['async']);
};
```

From my extension 

```php
class AcMarcheMercrediExtension extends Extension implements PrependExtensionInterface
{
    public function load(array $configs, ContainerBuilder $containerBuilder): void
    {
        $phpFileLoader = new PhpFileLoader($containerBuilder, new FileLocator(__DIR__.'/../../config'));
        $phpFileLoader->load('services.php');
        $phpFileLoader->load('packages/messenger.php'); //doesn't work     
    }
```
I have this error :

Could not resolve argument "Symfony\Config\FrameworkConfig $framework" for "/var/www/appli/src/AcMarche/Mercredi/src/DependencyInjection/../../config/packages/messenger.php".

**Comment by ro0NL** [comment]
what's the previous exception?

https://github.com/symfony/symfony/blob/e34cd7dd2c6d0b30d24cad443b8f964daa841d71/src/Symfony/Component/DependencyInjection/Loader/PhpFileLoader.php#L138-L140

**Comment by jfsenechal** [comment, marked as answer]
LogicException
You cannot use the ConfigBuilders without providing a class implementing ConfigBuilderGeneratorInterface.

You can see my bundle here https://github.com/acmarche/mercredi/blob/master/src/DependencyInjection/AcMarcheMercrediExtension.php

**Reply by ro0NL** [reply]
see https://github.com/symfony/symfony/blob/76a414c7d8ad03b5d6e520edfd8b947af21717b0/src/Symfony/Component/HttpKernel/Kernel.php#L765 how core initializes PhpFileLoader

not sure what can be improved out-of-the-box. A feature request could be to abstract the loader factorization for you (eg. $this->load('file.php')` in extension, and done :)

**Reply by jfsenechal** [reply]
when placing a file like "messenger.php" in the config folder of the application, how does symfony load it ?

**Reply by ro0NL** [reply]
using the DelegatingLoader as linked above

**Reply by jfsenechal** [reply]
Like that ?
```
       
        $resolver = new LoaderResolver([
            $phpFileLoader,
            new ClosureLoader($containerBuilder),
        ]);

        $delegatingLoader = new DelegatingLoader($resolver);
        $delegatingLoader->load('packages/messenger.php');
```
I have the same error Could not resolve argument "Symfony\Config\FrameworkConfig $framework" for "/var/www/mercredi5/src/AcMarche/Mercredi/src/DependencyInjection/../../config/packages/messenger.php".

**Reply by ro0NL** [reply]
the kernel uses its own loader, as linked above: 

https://github.com/symfony/symfony/blob/76a414c7d8ad03b5d6e520edfd8b947af21717b0/src/Symfony/Component/HttpKernel/Kernel.php#L757-L772


here we see php maps to `new PhpFileLoader($container, $locator, $env, class_exists(ConfigBuilderGenerator::class) ? new ConfigBuilderGenerator($this->getBuildDir()) : null),` where a ConfigBuilderGenerator is passed, which solves the error.

In your case another loader is created by you: `$phpFileLoader = new PhpFileLoader($containerBuilder, new FileLocator(__DIR__.'/../../config'));` where no ConfigBuilderGenerator is passed, which causes the error.

**Reply by jfsenechal** [reply]
Yes it works :-)

Thank you :-)

```php
public function load(array $configs, ContainerBuilder $containerBuilder): void
    {
        $this->loader->load('services.php');
        $this->loader->load('packages/messenger.php');
}
```
```php
protected function initPhpFilerLoader(ContainerBuilder $containerBuilder): PhpFileLoader
    {
        return new PhpFileLoader(
            $containerBuilder,
            new FileLocator(__DIR__.'/../../config/'),
            null,
            class_exists(ConfigBuilderGenerator::class) ? new ConfigBuilderGenerator(
                $containerBuilder->getParameter('kernel.cache_dir')
            ) : null
        );
    }
```

**Reply by ro0NL** [reply]
nice. I'd fill in the environment parameter as well (getParameter(kernel.environment). IIUC it enables compatibility with certain features like https://symfony.com/blog/new-in-symfony-5-3-configure-multiple-environments-in-a-single-file

cc @Nyholm WDYT about improving this?
